### PR TITLE
Add TosaBuilder for helper functions that require rewriter 

### PR DESCRIPTION
**Pull request by maxbartel**
This adds the `TosaBuilder `with const, reshape and transpose. I haven't deleted the old helper functions because I only transitioned some ops to the builder class to get feedback.

### CODE REVIEW

File: src/Conversion/ONNXToTOSA/DialectBuilder.hpp [hunk 1]
```diff
@@ -30,6 +30,30 @@ namespace onnx_mlir {
...
+  TosaBuilder(mlir::Location loc) : DialectBuilder(loc) {}
+  TosaBuilder(mlir::PatternRewriter &b, mlir::Location loc)
+      : DialectBuilder(b, loc), patternRewriter(&b) {}
+  TosaBuilder(const DialectBuilder &db) : DialectBuilder(db) {}
+  virtual ~TosaBuilder() {}
+
+  mlir::Value reshape(mlir::Value &value, llvm::ArrayRef<int64_t> shape);
+  mlir::Value transpose(mlir::Value &value, llvm::ArrayRef<int64_t> perm);
+  template <typename T>
+  mlir::Value getConst(llvm::ArrayRef<T> vec, llvm::ArrayRef<int64_t> shape);
```
Comment (ljfitz) [inline src/Conversion/ONNXToTOSA/DialectBuilder.hpp:43]: Is this inspired by some existing implementation? Comparing the general implementation and the specialization for `float` I'm a little concerned that the type assumptions are incorrectly overgeneralized.

If this is indeed an over-generalization, maybe a straightforward overloading in the header file for the supported types would make more sense, and if the implementations can share code via a template then do that in the cpp file?

**Review by ljfitz — APPROVED** [review]
The call sites do look easier to read, and can see this can be nice for a reasonably compact target dialect where the class is unlikely to get overly bloated.

File: src/Conversion/ONNXToTOSA/DialectBuilder.hpp (hunk 1 above)
Comment (maxbartel) [reply src/Conversion/ONNXToTOSA/DialectBuilder.hpp:43]: These are the functions from [Ferdinand](https://github.com/ljfitz/onnx-mlir/blob/fcfca0a2cdd358ef04f15ae4804b2bd858374211/src/Conversion/ONNXToTOSA/ONNXToTOSALegalizeUtils.cpp#L90-L142) that I just moved into the class. I think they are originally from TF.

**Comment by maxbartel** [comment]
I only plan to move the helper functions that create ops to this class and maybe one or two obvious choices like the `reshapeOp `that can take a normal `int64_t `vector as a shape

**Comment by maxbartel** [comment]
I will finish this up and then re-request your review for the full move to this class.

**Comment by maxbartel** [comment]
@ljfitz It seems like` Gemm.cpp` was hit by clang-format. Sorry for that!

File: test/mlir/conversion/onnx_to_tosa/NN/MaxPoolSingleOut.mlir [hunk 2]
```diff
@@ -9,8 +9,8 @@ func.func @test_default_maxpoolsingleout(%arg0 : tensor<5x5x32x32xf32>) -> tenso
 // CHECK-DAG:   "tosa.const"() {value = dense<[0, 2, 3, 1]> : tensor<4xi64>} : () -> tensor<4xi64>
 // CHECK-DAG:   %[[TRANS_ARG:.*]] = "tosa.transpose"(%arg0, %0) : (tensor<5x5x32x32xf32>, tensor<4xi64>) -> tensor<5x32x32x5xf32>
 // CHECK-DAG:   %[[MPOOL_RES:.*]] = "tosa.max_pool2d"(%[[TRANS_ARG]]) {kernel = [3, 3], pad = [0, 0, 0, 0], stride = [1, 1]} : (tensor<5x32x32x5xf32>) -> tensor<5x30x30x5xf32>
-// CHECK-DAG:   "tosa.const"() {value = dense<[0, 3, 1, 2]> : tensor<4xi32>} : () -> tensor<4xi32>
-// CHECK-DAG:   %[[TRANS_MPOOL_RES:.*]] = "tosa.transpose"(%[[MPOOL_RES]], %3) : (tensor<5x30x30x5xf32>, tensor<4xi32>) -> tensor<5x5x30x30xf32>
+// CHECK-DAG:   "tosa.const"() {value = dense<[0, 3, 1, 2]> : tensor<4xi64>} : () -> tensor<4xi64>
+// CHECK-DAG:   %[[TRANS_MPOOL_RES:.*]] = "tosa.transpose"(%[[MPOOL_RES]], %3) : (tensor<5x30x30x5xf32>, tensor<4xi64>) -> tensor<5x5x30x30xf32>
```
Comment (ljfitz) [inline test/mlir/conversion/onnx_to_tosa/NN/MaxPoolSingleOut.mlir:13]: In TOSA they must be int32_t. Is tosa-mlir more relaxed?

File: test/mlir/conversion/onnx_to_tosa/NN/MaxPoolSingleOut.mlir [hunk 3]
```diff
@@ -112,6 +112,6 @@ func.func @test_default_maxpoolsingleout_same_upper_ceil_mode(%arg0 : tensor<5x5
 // CHECK-DAG:     "tosa.const"() {value = dense<[0, 2, 3, 1]> : tensor<4xi64>} : () -> tensor<4xi64>
 // CHECK-DAG:     %[[TRANS_ARG:.*]] = "tosa.transpose"(%arg0, %0) : (tensor<5x5x16x13xf32>, tensor<4xi64>) -> tensor<5x16x13x5xf32>
 // CHECK-DAG:     %[[MPOOL_RES:.*]] = "tosa.max_pool2d"(%[[TRANS_ARG]]) {kernel = [4, 4], pad = [0, 0, 1, 2], stride = [4, 4]} : (tensor<5x16x13x5xf32>) -> tensor<5x4x4x5xf32>
-// CHECK-DAG:     "tosa.const"() {value = dense<[0, 3, 1, 2]> : tensor<4xi32>} : () -> tensor<4xi32>
-// CHECK-DAG:     %[[TRANS_MPOOL_RES:.*]] = "tosa.transpose"(%[[MPOOL_RES]], %3) : (tensor<5x4x4x5xf32>, tensor<4xi32>) -> tensor<5x5x4x4xf32>
+// CHECK-DAG:     "tosa.const"() {value = dense<[0, 3, 1, 2]> : tensor<4xi64>} : () -> tensor<4xi64>
+// CHECK-DAG:     %[[TRANS_MPOOL_RES:.*]] = "tosa.transpose"(%[[MPOOL_RES]], %3) : (tensor<5x4x4x5xf32>, tensor<4xi64>) -> tensor<5x5x4x4xf32>
```
Comment (ljfitz) [inline test/mlir/conversion/onnx_to_tosa/NN/MaxPoolSingleOut.mlir:116]: I see unexpected submodule updates in this pr.

File: test/mlir/conversion/onnx_to_tosa/NN/MaxPoolSingleOut.mlir (hunk 2 above)
Comment (maxbartel) [reply test/mlir/conversion/onnx_to_tosa/NN/MaxPoolSingleOut.mlir:13]: Yes, it is. I can also change this to `int32_t ` to match the standard

**Pull request merged**